datafusion/expr/src/function.rs

```diff
@@ -83,6 +83,9 @@ pub struct AccumulatorArgs<'a> {
     /// The input type of the aggregate function.
     pub input_type: &'a DataType,
 
+    /// If the input type is nullable.
+    pub input_nullable: bool,
```

Comment:
I think we only need `nullable` for state_field 🤔

Reply:
Yes, this is not used in the PR. I just thought it made sense to make the API more similar but will revert.


But I also noticed this is not enough to resolved the `array_agg` regression. There are two more limitiations in the current UDAF API. Firstly here
https://github.com/eejbyfeldt/datafusion/blob/18042fd69138e19613844580408a71a200ea6caa/datafusion/physical-expr-common/src/aggregate/mod.rs#L287-L289
the nullability of the returned field is hardcoded to `true` and it not controllable AggregateUDFImpl.  What is the desired way to fix this?

Should be api be changed to instead implement a method `fn field`? 

Or should we add a method `return_nullable` method with a default `false` implementation?


I also noticed that the current implementation for `array_agg` does not propagate the nullability of the input to the field in the returned array. This is probably because the `return_type` method does not have access to nullability. But probably something we want to be able to resolve in the long run.

Reply:
I think we can use `nullable` for `field()`.
```rust
    fn field(&self) -> Result<Field> {
        Ok(Field::new(&self.name, self.data_type.clone(), self.nullable))
    }
```


> I also noticed that the current implementation for array_agg does not propagate the nullability of the input to the field in the returned array. This is probably because the return_type method does not have access to nullability. But probably something we want to be able to resolve in the long run.

I think `nullable` is both set in `state_field` and `field`, so the returned array should match the schema of them. 🤔

Reply:
> I think we can use `nullable` for `field()`.

Looking closer at the code I see that this will maintain the behavior of the old code. But it seems wrong to me that we in general assume that the aggregate maintains the nullability of the input type. If we consider the aggregate `array_agg`. Then there are two "nullable" fields in the return value the "top level" value and the "field inside" the returned array. I think our `array_agg` (or at least a possible `array_agg`) will return an empty array when there are no values. This means that the nullability of the "top level" field should always be `false` regardless of input nullability and the nullabillity that depends on the input is the "field inside" the array.

Reply:
> This means that the nullability of the "top level" field should always be false

Why do we expect `not null` if there is possible empty array?

The nullability is introduced in https://github.com/apache/datafusion/pull/8055

I'm thinking maybe we could set it to always true 🤔 ?

Reply:
The nullability is introduced in https://github.com/apache/datafusion/pull/8055
There might be other way to fix #8055  🤔

Reply:
I test the code in #8032, and it there is no error after I change the field back to `false` 🤔

Reply:
Ideally the field is

```rust
 fn field(&self) -> Result<Field> {
        Ok(Field::new_list(
            &self.name,
            // This should be the same as return type of AggregateFunction::ArrayAgg
            Field::new("item", self.input_data_type.clone(), self.nullable),
            false,
        ))
    }
```

Reply:
Yes, that was what I was trying to explain. I created this PR that fixes that https://github.com/apache/datafusion/pull/11093 it required some other changes to make that change possible.